### PR TITLE
Simplify BlacklistingChannel

### PR DESCRIPTION
**Pull request by carterkozak**
More thread safety, use compare-and-swap where possible, reuse the instance. This change prevents us from mutating state based on old permits.

==COMMIT_MSG==
Simplify BlacklistingChannel
==COMMIT_MSG==

## Possible downsides?
Could introduce bugs.

### CODE REVIEW

File: dialogue-core/src/main/java/com/palantir/dialogue/core/BlacklistingChannel.java [hunk 1]
```diff
@@ -166,43 +136,59 @@ public String toString() {
 
     private final class BlacklistState {
```
Comment (ferozco) [inline dialogue-core/src/main/java/com/palantir/dialogue/core/BlacklistingChannel.java:113]: implements `LimitedChannel`

File: dialogue-core/src/main/java/com/palantir/dialogue/core/BlacklistingChannel.java [hunk 2]
```diff
@@ -70,6 +70,30 @@
...
+
+        @Override
+        public void onSuccess(Response response) {
+            // TODO(jellis): use the Retry-After header (if present) to determine how long to blacklist the channel
+            if (Responses.isQosStatus(response) || Responses.isServerError(response)) {
+                log.debug(
+                        "Blacklisting {} due to status code {}",
+                        UnsafeArg.of("delegate", delegate),
+                        SafeArg.of("code", response.code()));
+                channelBlacklistState.set(new BlacklistState(duration, NUM_PROBATION_REQUESTS));
```
Comment (ferozco) [inline dialogue-core/src/main/java/com/palantir/dialogue/core/BlacklistingChannel.java]: In cases where there are multiple requests in flight and they all fail, we will rest the blacklist multiple times. Should we CAS to start blacklisting on the first failure?

Comment (carterkozak) [reply dialogue-core/src/main/java/com/palantir/dialogue/core/BlacklistingChannel.java]: I went back and forth on this, I think you're right though and we should avoid resetting the timeout when requests come in.

File: dialogue-core/src/main/java/com/palantir/dialogue/core/BlacklistingChannel.java (hunk 1 above)
Comment (carterkozak) [reply dialogue-core/src/main/java/com/palantir/dialogue/core/BlacklistingChannel.java:113]: I'm not sure that buys us anything, does it?

File: dialogue-core/src/main/java/com/palantir/dialogue/core/BlacklistingChannel.java [hunk 3]
```diff
@@ -166,43 +136,59 @@ public String toString() {
...
+                return delegate.maybeExecute(endpoint, request)
+                        .map(future -> DialogueFutures.addDirectCallback(future, blacklistingCallback));
+            } else {
+                log.debug("Probation channel request not allowed");
+                return Optional.empty();
+            }
         }
 
-        BlacklistStage maybeProgressAndGet() {
+        Optional<Probation> maybeProgressAndGet() {
```
Comment (iamdanfox) [inline dialogue-core/src/main/java/com/palantir/dialogue/core/BlacklistingChannel.java]: rename to `isBlacklistingComplete` ? or `maybeBeginProbation`?

File: dialogue-core/src/main/java/com/palantir/dialogue/core/BlacklistingChannel.java [hunk 4]
```diff
@@ -166,43 +136,59 @@ public String toString() {
...
             if (inProbation.get()) {
-                return probation;
+                return Optional.of(probation);
             }
-            if (ticker.read() >= blacklistUntil.untilNanos) {
+            if (ticker.read() >= blacklistUntilNanos) {
                 if (inProbation.compareAndSet(false, true)) {
```
Comment (iamdanfox) [inline dialogue-core/src/main/java/com/palantir/dialogue/core/BlacklistingChannel.java]: elsewhere we have debug logs for interesting transitions - can we include one here for beginning probation?

File: dialogue-core/src/main/java/com/palantir/dialogue/core/BlacklistingChannel.java [hunk 5]
```diff
@@ -166,43 +136,59 @@ public String toString() {
 
     private final class BlacklistState {
         private final AtomicBoolean inProbation = new AtomicBoolean();
```
Comment (iamdanfox) [inline dialogue-core/src/main/java/com/palantir/dialogue/core/BlacklistingChannel.java]: why do we need a boolean and the `Probation` field - wouldn't an AtomicReference<Probation> or maybe volatile Probation do the trick using null or Optional.empty to signify that we're not in probation?

File: dialogue-core/src/main/java/com/palantir/dialogue/core/BlacklistingChannel.java (hunk 4 above)
Comment (carterkozak) [reply dialogue-core/src/main/java/com/palantir/dialogue/core/BlacklistingChannel.java]: good call

File: dialogue-core/src/main/java/com/palantir/dialogue/core/BlacklistingChannel.java (hunk 5 above)
Comment (carterkozak) [reply dialogue-core/src/main/java/com/palantir/dialogue/core/BlacklistingChannel.java]: Updated to use an AtomicReference for probation

File: dialogue-core/src/main/java/com/palantir/dialogue/core/BlacklistingChannel.java (hunk 3 above)
Comment (carterkozak) [reply dialogue-core/src/main/java/com/palantir/dialogue/core/BlacklistingChannel.java]: 👍

File: dialogue-core/src/main/java/com/palantir/dialogue/core/BlacklistingChannel.java (hunk 2 above)
Comment (carterkozak) [reply dialogue-core/src/main/java/com/palantir/dialogue/core/BlacklistingChannel.java]: Updated, it makes some of the simulations better :-)

**Review by iamdanfox — APPROVED** [review]

**Pull request merged**